docs/metatable-type-functions.md

```diff
@@ -0,0 +1,79 @@
+# Metatable Type Functions
+
+## Summary
+
+Implement type functions for `getmetatable` and `setmetatable`.
+
+## Motivation
+
+### `setmetatable` Type Function
+
+There is currently no way for users to apply metatable type information to a type without the usage of `typeof()`. This isn't ideal, as it adds verbosity and boilerplate to common patterns such as object-oriented programming. For example, the following:
+
+```lua
+local clock = {}
+type Identity = typeof(setmetatable({} :: { time: number }, { __index = clock }))
+```
+
+could be reduced to:
+
+```lua
+local clock = {}
+type Identity = setmetatable<{ time: number }, { __index: clock }>
+```
+
+### `getmetatable` Type Function
+
+[Issue #1435](https://github.com/luau-lang/luau/issues/1435) is caused by a lack of special behavior for `getmetatable` when the type's metatable has a `__metatable` field. This is fixable in a variety of ways, however a type function has multiple benefits and exposes more type-level expression for users. It fits in nicely with the new solver, as type functions are a cleaner alternative to magic functions.
+
+## Design
+
+### `setmetatable` Type Function
+
+In the following code example, `Identity` should evaluate to `{ sound: string, @metatable: { __index: animal } }`:
+
+```lua
+local animal = {}
+type Identity = setmetatable<{
+	sound: string
+}, {
+	__index: animal
+}>
+```
+
+### `getmetatable` Type Function
+
+In the following code example, `ReversedIdentity` should evaluate to `{ __index: animal }`:
+
+```lua
+local animal = {}
+type Identity = setmetatable<{
+	sound: string
+}, {
+	__index: animal
```

Comment:
Do you mean for this to be `typeof(animal)`? Or for `animal` to be a type?

Reply:
Yeah, I did. Oops